### ISSUE TITLE
Make time resource versions more predictable

### ISSUE THREAD

**Issue by pivotal-jwinters** [concourse/time-resource#45]
## What challenge are you facing?

Right now time resource versions are unpredictable. The resource `check` will output the current timestamp after its `interval` has elapsed (or whatever other configuration is used). However, if the `check` gets delayed then the gap between versions is unpredictable.

## What would make this better?

Make time resource versions more predictable. Instead of emitting the current timestamp, emit versions at set times based on the interval specified. This means that if the last version emitted was `1:42:05` and the interval was `1m` then if the check runs at say `1:45:30` the following versions would be emitted: `1:43:05`, `1:44:05`, `1:45:05`. This way regardless of when the check actually runs, we end up with a deterministic set of versions.

**Comment by owenfarrell** [comment]
So, I started hacking away at this yesterday, but came across some key questions. Below are the assumptions that I've started running with.

**A1:  When provided with a previous version, `check` should a list of all valid versions since the specified version**

The assumption here is to strictly adhere to the [contract](https://concourse-ci.org/implementing-resource-types.html#resource-check) for implementing a resource and that anyone who wants to trigger jobs for every version can configure the pipeline to do just that.

Example:
```
resources:
- name: this_resource_should_generate_3_new_versions_each_time_it_is_checked
  type: time
  check_every: 15m
  source:
    interval: 5m
```

**A2:  When defining an `interval` and a `start`, new versions will be generated based on the start time (not the previous version)**

The assumption here is to essentially "reset" the interval calculation each day. This ensures consistency for situations where the specified interval doesn't divide evenly in to a 24 hour period or where there aren't 24 hours in a day (switching to/from daylight saving time).

Example:
```
resources:
- name: this_interval_does_not_divide_evenly_in_to_24h
  type: time
  source:
    interval: 35m
    start: 3:00am
    stop: 6:00am
```

The above configuration would generate the following versions every day (note that these are the versions generated, not _**when**_ the versions are generated).

* 3:00:00
* 3:35:00
* 4:05:00
* 4:40:00
* 5:15:00
* 5:50:00

**A3: Implementing this kind of consistency only applies when an `interval` is defined.**

If an `interval` is not defined, then the existing behavior (generating a version based on the current timestamp) will continue when within the `start`/`stop` window.

**Comment by vito** [comment]
@owenfarrell Awesome - happy to see this being worked on!

A1 and A2 make sense to me.

For A3, to be honest I'm having a hard time interpreting what `start` + `stop` *without* `interval` should even mean. If it just returns the current time, that means the version history would be coupled to the cluster-wide checking interval, which seems like a bad idea. Does anyone use this functionality? Maybe we should just have it return the `start` time and ignore `stop`? 🤔 (Or just make `interval` required?)

**Comment by owenfarrell** [comment]
**tl;dr** I think there are legitimate use cases for the existing validation logic (`interval` and/or `stop` + `start`). But I also see the rationale for making `stop` optional.

Our use case was specifically to trigger infrequent jobs that don't have strict timing requirements (3rd party repository checks, key rotations, etc.).

> If it just returns the current time, that means the version history would be coupled to the cluster-wide checking interval

True story. In our scenario, we care that a new version is generated, but we don't particularly care what the version actually is.

Given the interval for our jobs, we don't want to burden Concourse with the management/overhead of running `check` on a frequent cadence when new versions are rarely (relatively speaking) generated.

So our approach relies on `check_every` + `start`/`stop` (but no `interval`) to generate exactly 1 new version at some point in the `start`/`stop` window.

**Comment by vito** [comment]
@owenfarrell Gotcha. In that case having it just return the start time (once it's elapsed) sounds reasonable to me. Though maybe we should rename it to `at` or something since `stop` isn't really respected.

So there would be 3 use cases:

* `interval`: run on a fixed interval
* `interval` + `start` + `stop`: run on a fixed interval within the time range
* `at`: run only at the specified time

How does that sound?

**Comment by owenfarrell** [comment]
@vito That makes sense to me.

I have a draft PR out there which addresses the original report (but not `at`). The logic looks a bit gnarly, but mostly because it's not DRY yet.

Any initial feedback would be appreciated!

**Comment by Kump3r** [comment]
Hello all,

We are also facing the same issue.
I am just checking in, that this is also relevant for us.
Thank you for the efforts!

**Comment by taylorsilva** [comment]
Looks like no one is working on this right now. If anyone wants to take a go at improving this, leave a comment and open a PR 🙏

**Comment by Kump3r** [comment]
I can try to, haven’t had time to check code, but I have a basic understanding of the desired result and problem reported :)

**Comment by Kump3r** [comment]
Looks like there was a PR #51, which was closed as #52, #53 , #54 addressed it, maybe the issue was just forgotten, or am I. missing something?

**Comment by Kump3r** [comment]
Even setting it to every minute, wouldn’t guarantee that the new version would be “predictable” as the one check might be scheduled a couple of miliseconds after the other and you wouldn’t still be able to predict. Also an observation here, seems that sometimes when you have the interval set to a minute, sometimes it takes over a minute for the check to be scheduled causing your versions to be 00:01:00 -> 00:03:05 -> 00:05:20 (hh:mm:ss) so if we do not take the timestamp, but rather the interval, at a certain point your job could have the fifth version, which is 00:05:00 (with the new implementation) but the actual job would be triggered in 00:10:03 for example. This could mislead and confuse the Concourse operator imho. Let me know what you think, but I am a but held back and have some concerns here. Maybe a source configuration, where the user, actively acknowledges the potential risk by using it, however still it will cause double work and not sure of the added value. Perhaps a deeper discussion is necessary to see if it actually brings merit.

**Comment by taylorsilva** [comment]
I agree with you. This whole thing honestly feels a bit impossible to handle given the way resources are run within Concourse. I'm struggling to see how we can make this better 😕
